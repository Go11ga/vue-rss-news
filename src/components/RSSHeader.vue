<template>
    <div class="rss-header">

        <div class="rss-header__item">
            <h1 class="rss-header__text">
                Список новостей
            </h1>

            <button class="rss-header__refresh" @click="fetchNews">
                <img class="rss-header__img" src="@/assets/style/images/refresh.png" alt="refresh">
            </button>
        </div>

        <div class="rss-header__item">
            <div class="rss-header__input-container">
                <input 
                    class="rss-header__input" 
                    v-model.trim="searchValue" 
                    @keyup.enter="searchNews"
                >
                <button class="rss-header__btn" @click="searchNews">
                    <img src="@/assets/style/images/search.png" alt="search">
                </button>
            </div>
        </div>

    </div>
</template>

<script>
import { mapMutations, mapActions } from 'vuex'

export default {
    name: 'RSSHeader',

    data: () => ({
        searchValue: ''
    }),

    methods: {
        ...mapMutations([
            'setSearchValue'
        ]),

        ...mapActions([
            'fetchNews'
        ]),

        searchNews() {
            this.setSearchValue(this.searchValue)
        }
    }
}
</script>
