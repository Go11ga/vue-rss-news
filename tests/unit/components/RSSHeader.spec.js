import { shallowMount, createLocalVue } from '@vue/test-utils'
import RSSHeader from '@/components/RSSHeader'
import Vuex from 'vuex'

describe('RSSHeader.vue рендеринг', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(RSSHeader)
    })

    afterEach(() => {
        wrapper.destroy()
    })

    it('правильно рендерится', () => {
        expect(wrapper.vm.$options.name).toMatch('RSSHeader')

        expect(wrapper.find('h1').text()).toMatch('Список новостей')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })
})

const localVue = createLocalVue()
localVue.use(Vuex)

describe('RSSHeader.vue пользовательские события', () => {

    let wrapper
    let mutations
    let actions
    let store
    let searchNewsSpy

    beforeEach(() => {
        mutations = {
            setSearchValue: jest.fn()
        }

        actions = {
            fetchNews: jest.fn()
        }
        
        store = new Vuex.Store({
            mutations,
            actions
        })

        searchNewsSpy = jest.spyOn(RSSHeader.methods, 'searchNews')

        wrapper = shallowMount(RSSHeader, {
            localVue,
            store
        })
    })

    afterEach(() => {
        wrapper.destroy()
    })

    it('клик на кнопку .rss-header__refresh вызывает Vuex action fetchNews', () => {
        wrapper.find('.rss-header__refresh').trigger('click')

        expect(actions.fetchNews).toHaveBeenCalled()
        expect(actions.fetchNews).toHaveBeenCalledTimes(1)
    })

    it('нажатие enter в инпуте вызывает метод searchNews', () => {
        let input = wrapper.find('.rss-header__input')
        
        input.trigger('keyup')
        expect(searchNewsSpy).not.toHaveBeenCalled()

        input.trigger('keyup.enter')
        
        expect(searchNewsSpy).toHaveBeenCalled()
        expect(searchNewsSpy).toHaveBeenCalledTimes(1)
    })

    it('клик на кнопку .rss-header__btn вызывает метод searchNews', () => {
        wrapper.find('.rss-header__btn').trigger('click')

        expect(searchNewsSpy).toHaveBeenCalled()
        expect(searchNewsSpy).toHaveBeenCalledTimes(2)
    })

    
    it('метод searchNews вызывает Vuex mutation setSearchValue с payload', () => {
        wrapper.setData({ searchValue: 'test' })
        wrapper.vm.searchNews()

        expect(mutations.setSearchValue).toHaveBeenCalled()
        expect(mutations.setSearchValue).toHaveBeenCalledTimes(1)
        expect(mutations.setSearchValue).toHaveBeenCalledWith({}, 'test')
    })
})
