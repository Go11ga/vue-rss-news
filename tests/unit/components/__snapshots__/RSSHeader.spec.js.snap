// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RSSHeader.vue рендеринг snapshot 1`] = `
<div
  class="rss-header"
>
  <div
    class="rss-header__item"
  >
    <h1
      class="rss-header__text"
    >
      
                Список новостей
            
    </h1>
     
    <button
      class="rss-header__refresh"
    >
      <img
        alt="refresh"
        class="rss-header__img"
        src="@/assets/style/images/refresh.png"
      />
    </button>
  </div>
   
  <div
    class="rss-header__item"
  >
    <div
      class="rss-header__input-container"
    >
      <input
        class="rss-header__input"
      />
       
      <button
        class="rss-header__btn"
      >
        <img
          alt="search"
          src="@/assets/style/images/search.png"
        />
      </button>
    </div>
  </div>
</div>
`;
